src/daemon/distribution_test.c

```diff
@@ -0,0 +1,1014 @@
+    double size;
+    int want_err;
+    double *want_get;
+  } cases[] = {
+      {
+          .num_buckets = 0,
+          .want_err = EINVAL,
+          .want_get = NULL,
+      },
+      {
+          .num_buckets = 10,
+          .size = -5,
+          .want_err = EINVAL,
+          .want_get = NULL,
+      },
+      {.num_buckets = 8, .size = 0, .want_err = EINVAL, .want_get = NULL},
+      {
+          .num_buckets = 10,
+          .size = 2,
+          .want_get = array_new_linear(10, 2),
+      },
+      {
+          .num_buckets = 20,
+          .size = 0.67,
+          .want_get = array_new_linear(20, 0.67),
+      },
+      {
+          .num_buckets = 48,
+          .size = 8.259,
+          .want_get = array_new_linear(48, 8.259),
+      }};
+
+  for (size_t i = 0; i < (sizeof(cases) / sizeof(cases[0])); ++i) {
+    printf("## Case %zu:\n", i);
+
+    distribution_t *d =
+        distribution_new_linear(cases[i].num_buckets, cases[i].size);
+
+    if (cases[i].want_get == NULL) {
+      EXPECT_EQ_PTR(cases[i].want_get, d);
+
+      if (cases[i].want_err != 0) {
+        EXPECT_EQ_INT(cases[i].want_err, errno);
+      }
+    } else {
+      CHECK_NOT_NULL(d);
+      EXPECT_EQ_UINT64(cases[i].num_buckets, distribution_get_num_buckets(d));
+
+      double *boundaries = distribution_get_buckets_boundaries(d);
+      uint64_t *counters = distribution_get_buckets_counters(d);
+
+      for (size_t j = 0; j < cases[i].num_buckets; ++j) {
+        EXPECT_EQ_DOUBLE(cases[i].want_get[j], boundaries[j]);
+        EXPECT_EQ_UINT64(0, counters[j]);
+      }
+
+      EXPECT_EQ_DOUBLE(0, distribution_get_sum_gauges(d));
+      free(boundaries);
+      free(counters);
+    }
+
+    free(cases[i].want_get);
+    distribution_destroy(d);
+  }
+
+  return 0;
+}
+
+DEF_TEST(distribution_new_exponential) {
+  struct {
+    size_t num_buckets;
+    double base;
+    double factor;
+    int want_err;
+    double *want_get;
+  } cases[] = {
+      {
+          .num_buckets = 0,
+          .want_err = EINVAL,
+          .want_get = NULL,
+      },
+      {
+          .num_buckets = 10,
+          .base = -52,
+          .want_err = EINVAL,
+          .want_get = NULL,
+      },
+      {.num_buckets = 8, .base = 0, .want_err = EINVAL, .want_get = NULL},
+      {.num_buckets = 12,
+       .base = 2,
+       .factor = 0,
+       .want_err = EINVAL,
+       .want_get = NULL},
+      {.num_buckets = 33,
+       .base = 7,
+       .factor = -5,
+       .want_err = EINVAL,
+       .want_get = NULL},
+      {
+          .num_buckets = 6,
+          .base = 2,
+          .factor = 3,
+          .want_get = array_new_exponential(6, 2, 3),
+      },
+      {
+          .num_buckets = 10,
+          .base = 5,
+          .factor = 6.75,
+          .want_get = array_new_exponential(10, 5, 6.75),
+      },
+      {
+          .num_buckets = 26,
+          .base = 1.01,
+          .factor = 4.64,
+          .want_get = array_new_exponential(26, 1.01, 4.64),
+      }};
+
```

Comment:
I'd add a corner case when 0 < base < 1. And when base = 1

Reply:
I will that, thank you!